src/ast/passes/semantic_analyser.cpp

```diff
@@ -3402,32 +3447,70 @@ void SemanticAnalyser::accept_statements(StatementList &stmts)
   }
 }
 
-void SemanticAnalyser::update_key_type(const Map &map, const MapKey &new_key)
+void SemanticAnalyser::update_key_type(const Map &map,
+                                       const MapKey &new_key,
+                                       bool is_tuple_literal)
 {
   if (const auto &key = map_key_.find(map.ident); key != map_key_.end()) {
     bool valid = true;
-    if (key->second.args_.size() == new_key.args_.size()) {
-      for (size_t i = 0; i < key->second.args_.size(); i++) {
-        SizedType &key_type = key->second.args_[i];
-        const SizedType &new_key_type = new_key.args_[i];
-        if (key_type.IsStringTy() && new_key_type.IsStringTy()) {
-          key_type.SetSize(
-              std::max(key_type.GetSize(), new_key_type.GetSize()));
-        } else if (key_type != new_key_type) {
+    bool replace_key = false;
+    if (key->second.arg_.GetTy() == new_key.arg_.GetTy()) {
+      if (key->second.arg_.IsStringTy()) {
+        key->second.arg_.SetSize(
+            std::max(key->second.arg_.GetSize(), new_key.arg_.GetSize()));
+      } else if (key->second.arg_.IsTupleTy()) {
+        auto &keyFields = key->second.arg_.GetFields();
+        auto &newKeyFields = new_key.arg_.GetFields();
+        if (keyFields.size() != newKeyFields.size()) {
           valid = false;
-          break;
+        } else {
+          for (size_t i = 0; i < keyFields.size(); i++) {
+            SizedType &key_type = keyFields[i].type;
+            const SizedType &new_key_type = newKeyFields[i].type;
+            if (key_type.GetTy() != new_key_type.GetTy()) {
+              valid = false;
+              break;
+              // We can't exchange the keys if we don't have access to the
+              // underlying expressions
+            } else if (key_type.IsStringTy() && is_tuple_literal) {
```

Comment:
Related to comment about all tuples being literal, I'm not sure why a literal check is necessary here.

Reply:
I guess the comment above this statement didn't make it clearer 😞 Only in the case of a tuple literal (not a variable that is a tuple type) can we get access to the underlying expressions and manipulate their types. This is for cases like this:
```
@[(1, "hi")] = 1;
@[(2, "hello")] = 2;
```
We have to increase the string size of "hi" and, because of tuple alignment (comment below), we just replace the whole stored tuple type with the larger one.

Reply:
Ah ok. I was confused with the literal check. with dyn_cast this all makes sense